src/async-wrap.cc

```diff
@@ -177,6 +179,136 @@ static void PushBackDestroyId(Environment* env, double id) {
+  }
+
+  if (async_hooks->fields()[AsyncHooks::kBefore] > 0) {
+    Local<Value> uid = Number::New(wrap->env()->isolate(), wrap->get_id());
+    Local<Function> fn = wrap->env()->async_hooks_before_function();
+    TryCatch try_catch(wrap->env()->isolate());
+    MaybeLocal<Value> ar = fn->Call(
+        wrap->env()->context(), Undefined(wrap->env()->isolate()), 1, &uid);
+    if (ar.IsEmpty()) {
+      ClearFatalExceptionHandlers(wrap->env());
+      FatalException(wrap->env()->isolate(), try_catch);
+      return false;
+    }
+  }
+  return true;
+}
+
+
+static bool PostCallbackExecution(AsyncWrap* wrap) {
+  AsyncHooks* async_hooks = wrap->env()->async_hooks();
+
+  // If the callback failed then the after() hooks will be called at the end
+  // of _fatalException().
+  if (async_hooks->fields()[AsyncHooks::kAfter] > 0) {
+    Local<Value> uid = Number::New(wrap->env()->isolate(), wrap->get_id());
+    Local<Function> fn = wrap->env()->async_hooks_after_function();
+    TryCatch try_catch(wrap->env()->isolate());
+    MaybeLocal<Value> ar = fn->Call(
+        wrap->env()->context(), Undefined(wrap->env()->isolate()), 1, &uid);
+    if (ar.IsEmpty()) {
+      ClearFatalExceptionHandlers(wrap->env());
+      FatalException(wrap->env()->isolate(), try_catch);
+      return false;
+    }
+  }
+
+  if (wrap->env()->using_domains()) {
+    Local<Value> domain_v = wrap->object()->Get(wrap->env()->domain_string());
+    if (domain_v->IsObject()) {
+      Local<Object> domain = domain_v.As<Object>();
+      if (domain->Get(wrap->env()->disposed_string())->IsTrue())
+        return false;
+      Local<Value> exit_v = domain->Get(wrap->env()->exit_string());
+      if (exit_v->IsFunction()) {
+        if (exit_v.As<Function>()->Call(domain, 0, nullptr).IsEmpty()) {
+          FatalError("node::AsyncWrap::MakeCallback",
+                    "domain exit callback threw, please report this");
+        }
+      }
+    }
+  }
+
+  return true;
+}
+
+class PromiseWrap : public AsyncWrap {
+ public:
+  PromiseWrap(Environment* env, Local<Object> object)
+    : AsyncWrap(env, object, PROVIDER_PROMISE) {}
+  size_t self_size() const override { return sizeof(*this); }
+};
+
+
+static void GetPromiseDomain(Local<v8::Name> name,
+                             const v8::PropertyCallbackInfo<v8::Value>& info) {
+  Local<Context> context = info.GetIsolate()->GetCurrentContext();
+  Environment* env = Environment::GetCurrent(context);
+  info.GetReturnValue().Set(
+      info.Data().As<Object>()->Get(context,
+                                    env->domain_string()).ToLocalChecked());
+}
+
+
+static void PromiseHook(PromiseHookType type, Local<Promise> promise,
+                        Local<Value> parent, void* arg) {
+  Local<Context> context = promise->CreationContext();
+  Environment* env = Environment::GetCurrent(context);
+  const char async_id_key[] = "__async_wrap";
+  const char tag_id_key[] = "__async_wrap_tag";
+  if (type == PromiseHookType::kInit) {
```

Comment:
@addaleax As far as I can tell, the line above is only used to avoid invoking init hooks if none have been registered. I was hoping to avoid attaching the new async hook related properties to promises in the case where async hooks were disabled. Adding these properties causes ~30x increase in the time it takes to construct/resolve promises (adding the property causes all of V8's map checks to fail and prevent promises from ever running on fast paths). I am hoping to fix this by adding an internal field to promises (https://codereview.chromium.org/2889863002/) but it would be good to avoid modifying promises if async hooks are not in use.

This slow down may also come in to play when domains are active as that also seems to modify promise shape but I haven't run any benchmarks against domains yet.